### ISSUE TITLE
Validate GraphQL queries in gatsby-node.js with Relay Compiler

### ISSUE THREAD

**Issue by 3axap4eHko** [gatsbyjs/gatsby#1689]
If I put in a page template graphql filter like:
```
filter: { frontmatter: { date: { ne: null } } }
```
it throws an error `Unknown ast kind: NullValue` But it works in the `gatsby-node.js` with `createPages` method

**Comment by KyleAMathews** [comment]
We use the Relay Compiler to validate queries in pages so we get more warnings there. Would like to do the same validation for queries written elsewhere e.g. in gatsby-node.js

**Comment by KyleAMathews** [comment]
Updated the issue title to reflect this :-)

**Comment by cr101** [comment]
Just sharing this announcement about a [new GraphQL tool](https://www.graph.cool/blog/2017-08-01-graphql-config-and-cli-aeghee3di9/)

**Comment by nkohari** [comment]
I ran into this problem as well when trying to get a list of all blog posts without the static pages ("about me", etc.) Here's the query:

```graphql
{
  allMarkdownRemark(filter:{ frontmatter: { date: { ne: null }}}) {
    edges {
      node {
        frontmatter {
          title
          date
        }
      }
    }
  }
}
```

This works as expected in GraphiQL, but returns the `Unknown ast kind: NullValue` error in the site itself. I'm not sure that this is a problem with validation, because the query seems to be correct. Is there another way to go about this?

**Comment by jquense** [comment]
I'm not sure actually where the `null` literal error is coming from. The `null` literal is a fairly new addition to the GQL language, so it's possible some place is using an older version of graphql, or the Relay compiler is intentionally protecting against it (which we should change)

**Comment by nkohari** [comment]
I think you're right. As best as I can tell, the error seems to be coming from the Relay compiler, from this block:

https://github.com/gatsbyjs/gatsby/blob/f8956a6449475c1dc364f328ab173857b4e9ebcd/packages/gatsby/src/internal-plugins/query-runner/query-compiler.js#L123-L131

Tracking it down in the Relay source, it looks like `RelayParser._transformValue()` is the culprit:

https://github.com/facebook/relay/blob/23864eeddc915f119f7b0c698a3d48ceab7488f9/packages/relay-compiler/graphql-compiler/core/RelayParser.js#L735

This switch statement doesn't contain a case for `NullValue`, but if I add one like this:

```js
...
switch (ast.kind) {
  case 'NullValue':
    return {
      kind: 'Literal',
      metadata: null,
      value: null
    };
  ...
}
```

The error no longer occurs, and the query runs as expected. I'm not very familiar with the inner-workings of GraphQL and Relay, but it seems like this is an issue in Relay instead, right?

**Comment by nickeblewis** [comment]
Will be keen to know how this pans out, as I tried the same thing today and having to work around it at the moment by ensuring my data has no null paths!! However there are occasions where I may wish to set them as null, for example, to signify that they are not "published"

**Comment by tim-field** [comment]
For those finding this outside of gatsby, the issues has been fixed and merged into relay and realy-compiler  ^1.3.0

**Comment by KyleAMathews** [comment]
Sweet! Thanks @nkohari for fixing the bug upstream!!
